flink-test-utils-parent/flink-test-utils/src/main/java/org/apache/flink/streaming/util/PseudoRandomValueSelector.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.streaming.util;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.annotation.VisibleForTesting;
+import org.apache.flink.configuration.ConfigOption;
+import org.apache.flink.configuration.Configuration;
+import org.apache.flink.runtime.util.EnvironmentInformation;
+
+import net.jcip.annotations.NotThreadSafe;
+import org.apache.commons.io.IOUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.charset.Charset;
+import java.util.Optional;
+import java.util.Random;
+import java.util.function.Function;
+
+/**
+ * Initializes the {@link Configuration} for particular {@link ConfigOption}s with random values if
+ * unset.
+ *
+ * <p>With the same seed, the same values are always selected if the {@link #select(Configuration,
+ * ConfigOption, Object[])} invocation happens in the same order. A different seed should select
+ * different values.
+ *
+ * <p>The seed is calculated from a global seed (~unique per build) and a seed specific to test
+ * cases. Thus, two different builds will mostly result in different values for the same test case.
+ * Similarly, two test cases in the same build will have different randomized values.
+ *
+ * <p>The seed can be set with the maven/system property test.randomization.seed and is set by
+ * default to commit id. If the seed is empty, {@link EnvironmentInformation} and as a last fallback
+ * git command is used to retrieve the commit id.
+ */
+@Internal
+@NotThreadSafe
```

Comment:
What part of this class is not thread safe?

Reply:
It doesn't adhere to the contract that it will produce the same randomized configuration for the same seeds if you start using it in parallel (the state is hidden in `Random` itself).